docs/understand/programming_model.rst

```diff
@@ -111,46 +160,213 @@ typically look the following:
+    through additional APIs.
+
+  Surface
+    A read-write version of texture memory.
+
+Execution model
+===============
+
+HIP programs consist of two distinct scopes:
+
+* The host-side API running on the host processor. There are to APIs available:
+
+  * The HIP runtime API which enables use of the single-source programming
+    model.
+
+  * The HIP driver API which sits at a lower level and most importantly differs
+    by removing some facilities provided by the runtime API, most
+    importantly around kernel launching and argument setting. It is geared
+    towards implementing abstractions atop, such as the runtime API itself.
+
+* The device-side kernels running on GPUs. Both the host and the device-side
+  APIs have synchronous and asynchronous functions in them.
+
+Host-side execution
+-------------------
+
+The part of the host-side API which deals with device management and their
+queries are synchronous. All asynchronous APIs, such as kernel execution, data
+movement and potentially data allocation/freeing all happen in the context of
+device streams.
+
+Streams are FIFO buffers of commands to execute relating to a given device.
+Commands which enqueue tasks on a stream all return promptly and the command is
+executed asynchronously. All side effects of a command on a stream are visible
+to all subsequent commands on the same stream. Multiple streams may point to
+the same device and those streams may be fed from multiple concurrent host-side
+threads. Execution on multiple streams may be concurrent but isn't required to
+be.
+
+Asynchronous APIs involving a stream all return a stream event which may be
+used to synchronize the execution of multiple streams. A user may enqueue a
+barrier onto a stream referencing an event. The barrier will block until
+the command related to the event does not complete, at which point all
+side effects of the command shall be visible to commands following the barrier,
+even if those side effects manifest on different devices.
+
+Streams also support executing user-defined functions as callbacks on the host.
+The stream will not launch subsequent commands until the callback completes.
+
+Device-side execution
+---------------------
+
+The SIMT programming model behind the HIP device-side execution is a
+middle-ground between SMT (Simultaneous Multi-Threading) programming known from
+multicore CPUs, and SIMD (Single Instruction, Multiple Data) programming
+mostly known from exploiting relevant instruction sets on CPUs (for example
+SSE/AVX/Neon).
+
+A HIP device compiler maps our SIMT code written in HIP C++ to an inherently
+SIMD architecture (like GPUs) not by exploiting data parallelism within a
+single instance of a kernel and spreading identical instructions over the SIMD
+engines at hand, but by scalarizing the entire kernel and issuing the scalar
+instructions of multiple kernel instances to each of the SIMD engine lanes.
```

Comment:
This is (almost) a copy of the first paragraph of the SIMT section.

Reply:
Removed.